runtime/jvmti/jvmtiClass.c

```diff
@@ -1198,6 +1198,18 @@ redefineClassesCommon(jvmtiEnv* env,
 		vm->internalVMFunctions->acquireExclusiveVMAccess(currentThread);
 	}
 
+#if defined(J9VM_JIT_CLASS_UNLOAD_RWMONITOR)
+	omrthread_rwmutex_enter_write(vm->classUnloadMutex);
+#else
+	omrthread_monitor_enter(vm->classUnloadMutex);
+#endif /* defined(J9VM_JIT_CLASS_UNLOAD_RWMONITOR) */
```

Comment:
Should `classUnloadMutex` only be acquired for `acquireSafePointVMAccess`? No other thread should run with exclusive VM access.

Reply:
I believe that safe-point VM access also prevents all other threads from holding VM access just as regular exclusive VM access does. The difference is that when acquiring safe-point VM access, we allow threads to run until they reach one of the reduced set of safe points. In particular, any JIT frame on top of the stack is allowed to keep running until it reaches a call, `monenter`, or `asynccheck` (or maybe `ResolveCHK`, or the stack check in the prologue if it's only just been called), or otherwise until it returns. This way we can be sure that once redefinition gets safe-point VM access, all JIT frames are paused at one of those points and not at some other GC point, e.g. an allocation. From the JIT compiler's perspective, this means that redefinition can only occur at those points, which helps to better optimize the code.

The whole point of this PR though is that taking exclusive VM access does *not* necessarily cause compilations to pause. Most of the time compilations don't hold VM access. If/when they need to inspect objects, they will take VM access for a short time and then release it again. Instead of holding VM access all the time, they hold the class unload mutex (non-exclusively). This prevents classes from being unloaded while compilations are looking at them even though those compilations don't hold VM access. In the same way, this PR uses the same class unload mutex to prevent classes from being redefined while compilations are looking at them.

Reply:
By the way, we have the (probably poorly exercised) option `-Xjit:disableNoVMAccess`, which looks to me like it causes compilations to hold VM access all the time. I imagine that's how everything used to work before the class unload mutex was implemented

Reply:
@gacholio Do you recommend `classUnloadMutex` be acquired before VM access (to prevent a deadlock) or does it not matter in this case due to exclusive VM access?

```
VM-access enter
classUnloadMutex enter

classUnloadMutex exit
VM-access exit

OR

classUnloadMutex enter
VM-access enter

VM-access exit
classUnloadMutex exit
```

Reply:
Just a note that with this change `redefineClassesCommon()` acquires and releases exclusive VM access and the class unload mutex in the same order as it did before. The class unload mutex is mostly just acquired a bit earlier (w.r.t. fixups). Previously it would be acquired and released within `jitClassesRedefined()` once all the other fixups were done, but that was still all while holding exclusive VM access. It's also acquired in slightly more cases: before, we'd skip acquiring the class unload mutex for an empty class list in fast HCR mode

Reply:
Also, from a quick look at the GC, I think class unloading also acquires the class unload mutex while already holding exclusive VM access. I looked at `MM_IncrementalGenerationalGC`, and there:
- `mainThreadGarbageCollect()` asserts that it has exclusive VM access and calls
- `runGlobalGarbageCollection()`, which calls
- `globalMarkPhase()`, which calls
- `postMarkMapCompletion()`, which calls
- `unloadDeadClassLoaders()`, which finally calls
- `MM_ClassLoaderManager::enterClassUnloadMutex()`.